cpp/src/arrow/compute/exec/aggregate_node.cc

```diff
@@ -132,12 +215,14 @@ class ScalarAggregateNode : public ExecNode {
 
     return plan->EmplaceNode<ScalarAggregateNode>(
         plan, std::move(inputs), schema(std::move(fields)), std::move(target_field_ids),
-        std::move(aggregates), std::move(kernels), std::move(states));
+        std::move(segment_field_ids), std::move(segmenter), std::move(aggregates),
+        std::move(kernels), std::move(states));
   }
 
   const char* kind_name() const override { return "ScalarAggregateNode"; }
 
   Status DoConsume(const ExecSpan& batch, size_t thread_index) {
+    GatedSharedLock lock(gated_shared_mutex_);
```

Comment:
Why is this lock needed?  Each batch of incoming data should be more or less separably processable right?  Or is this to handle the case where a segment spreads out across multiple batches?  Ideally we should not be holding any locks when we make kernel functions.

Reply:
Without these locks, I observed a race condition between the finishing of one segment, which reads accumulated states, and the consuming of the next segment, which accumulates states. The `GatedSharedLock` in the code ensure that the (multi-threaded) consuming of the next segment does not run until the finishing of the current segment completes. I can think of ways to reduce the locking, but they would be more complex and are better left for a future task.